profiler/src/test/java/com/splunk/opentelemetry/profiler/context/SpanContextualizerTest.java

```diff
@@ -0,0 +1,328 @@
+    assertNoLinkage(testClass, events);
+  }
+
+  @Test
+  void testNestedChildSpanOneThread() {
+    SpanContextualizer testClass = new SpanContextualizer();
+
+    String childSpanId = "xyz987";
+    String rawStack1 = "raw is raw1";
+    String rawStack2 = "raw is raw2";
+    long threadId = 906;
+
+    Events parent = buildEvents(spanId, threadId);
+    Events child = buildEvents(childSpanId, threadId);
+
+    testClass.updateContext(parent.scopeStart);
+    assertLinkage(testClass, parent, rawStack1);
+
+    testClass.updateContext(child.scopeStart);
+    assertLinkage(testClass, child, rawStack2);
+
+    testClass.updateContext(child.scopeEnd);
+    assertLinkage(testClass, parent, rawStack1);
+
+    testClass.updateContext(parent.scopeEnd);
+    assertNoLinkage(testClass, parent);
+  }
+
+  @Test
+  void testOneSpanThreadHops() {
+    SpanContextualizer testClass = new SpanContextualizer();
+
+    Events events1 = buildEvents(spanId, 906);
+    Events events2 = buildEvents(spanId, 907);
+
+    testClass.updateContext(events1.scopeStart); // has some scope on the first thread
+    testClass.updateContext(events2.scopeStart); // context switch to a new thread
+
+    assertLinkage(testClass, events1);
+    assertLinkage(testClass, events2);
+
+    testClass.updateContext(events2.scopeEnd);
+
+    assertNoLinkage(testClass, events2);
+    assertLinkage(testClass, events1); // thread 1 is still linked
+  }
+
+  @Test
+  void testMultipleThreadsOneSpanAllLinked() {
+    SpanContextualizer testClass = new SpanContextualizer();
+
+    long threadId1 = 907;
+    long threadId2 = 908;
+    long threadId3 = 909;
+    Events events1 = buildEvents(spanId, threadId1);
+    Events events2 = buildEvents(spanId, threadId2);
+    Events events3 = buildEvents(spanId, threadId3);
+
+    testClass.updateContext(events1.scopeStart); // context switch to a new thread
+    assertLinkage(testClass, events1);
+    assertNoLinkage(testClass, events2);
+    assertNoLinkage(testClass, events3);
+
+    testClass.updateContext(events2.scopeStart); // context switch to a new thread
+    assertLinkage(testClass, events1);
+    assertLinkage(testClass, events2);
+    assertNoLinkage(testClass, events3);
+
+    testClass.updateContext(events3.scopeStart); // context switch to a new thread
+    assertLinkage(testClass, events1);
+    assertLinkage(testClass, events2);
+    assertLinkage(testClass, events3);
+
+    testClass.updateContext(events2.scopeEnd);
+    assertLinkage(testClass, events1);
+    assertNoLinkage(testClass, events2);
+    assertLinkage(testClass, events3);
+
+    testClass.updateContext(events3.scopeEnd);
+    assertLinkage(testClass, events1);
+    assertNoLinkage(testClass, events2);
+    assertNoLinkage(testClass, events3);
+
+    testClass.updateContext(events1.scopeEnd);
+    assertNoLinkage(testClass, events1);
+    assertNoLinkage(testClass, events2);
+    assertNoLinkage(testClass, events3);
+  }
+
+  @Test
+  void testBackgroundWorkerThreads() {
```

Comment:
Isn't this scenario virtually the same as `testMultipleThreadsOneSpanAllLinked`?

Reply:
Very similar, but different completion order.